lnwallet/fee_estimator.go

```diff
@@ -458,3 +475,270 @@ func (b *BitcoindFeeEstimator) fetchEstimate(confTarget uint32) (SatPerKWeight,
 // A compile-time assertion to ensure that BitcoindFeeEstimator implements the
 // FeeEstimator interface.
 var _ FeeEstimator = (*BitcoindFeeEstimator)(nil)
+
+// WebAPIFeeEstimator is an implementation of the FeeEstimator interface that
+// queries an HTTP-based fee estimation from an existing web API.
+type WebAPIFeeEstimator struct {
+	started uint32 // To be used atomically.
+	stopped uint32 // To be used atomically.
+
+	// apiURL is the backing web API source we'll use for our queries.
+	apiURL string
+
+	// chainNotifier is used by the WebAPIFeeEstimator to determine when to
+	// re-query the API for fresh fees. We re-query whenever a new block
+	// comes in.
+	chainNotifier chainntnfs.ChainNotifier
+
+	// blockEpochs is our block notification subscription.
+	blockEpochs *chainntnfs.BlockEpochEvent
+
+	// feeByBlockTarget is our cache for fees pulled from the API. When a
+	// fee estimate request comes in, we pull the estimate from this array
+	// rather than re-querying the API, to prevent an inadvertent DoS attack.
+	feesMtx          sync.RWMutex
+	feeByBlockTarget [MaxBlockTarget + 1]uint32
+
+	// defaultFeePerKw is a fallback value that we'll use if we're unable
+	// to query the API for any reason.
+	defaultFeePerKw SatPerKWeight
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewWebAPIFeeEstimator creates a new WebAPIFeeEstimator from a given URL and a
+// fallback default fee. The fees are updated whenever a new block is mined.
+func NewWebAPIFeeEstimator(
+	api string, notifier chainntnfs.ChainNotifier, defaultFee SatPerKWeight) *WebAPIFeeEstimator {
+
+	return &WebAPIFeeEstimator{
+		apiURL:           api,
+		chainNotifier:    notifier,
+		feeByBlockTarget: [MaxBlockTarget + 1]uint32{},
+		defaultFeePerKw:  defaultFee,
+		quit:             make(chan struct{}),
+	}
+}
+
+// EstimateFeePerKW takes in a target for the number of blocks until an initial
+// confirmation and returns the estimated fee expressed in sat/kw.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) EstimateFeePerKW(numBlocks uint32) (SatPerKWeight, error) {
+	if numBlocks > MaxBlockTarget {
+		numBlocks = MaxBlockTarget
+	}
+
+	w.feesMtx.RLock()
+	feePerKb := w.feeByBlockTarget[numBlocks]
```

Comment:
handle the case where `numBlocks≈ is not found.

Reply:
we error if it isn't now